src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -2,24 +2,41 @@ package org.apache.spark.datasource
 
 import cats.Eval
 import com.codahale.metrics._
-import org.apache.spark._
-
 import java.util.concurrent.ConcurrentHashMap
+import org.apache.spark._
+import org.log4s._
+import scala.collection.JavaConverters._
 
 class MetricsSource {
   // Add metricNamespace to differentiate with spark system metrics.
   // Keeps track of all the Metric instances that are being published
-  val metricsMap = new ConcurrentHashMap[String, Eval[Counter]]
+  val metricsMap = new ConcurrentHashMap[String, Eval[SourceCounter]]
+
+  def getOrCreateCounter(metricNamespace: String, name: String): Counter = {
+    val ctx = Option(TaskContext.get())
+
+    def getNumberOrEmpty(getter: TaskContext => AnyVal): String =
+      ctx.map(getter).map(_.toString()).getOrElse("")
+
+    // The number of fields in the name should be consistent, even if there is no value for
+    // the attempt numbers
+    val stageId = getNumberOrEmpty(_.stageId())
+    val stageAttempt = getNumberOrEmpty(_.stageAttemptNumber())
+    val partitionId = getNumberOrEmpty(_.partitionId())
+    val taskAttempt = getNumberOrEmpty(_.taskAttemptId())
+
+    val metricName = s"$stageId.$stageAttempt.$partitionId.$taskAttempt.$name"
```

Comment:
On the cardinality front, this can make x200 (raw partitions), x3-5 (task attempts), x3-5 (stage attempts) -> x?? (num stages). And when it gets propagated to prometheus (in driver it does) it still impacts prometheus even after it stops being exported (prometheus still remembers names+labels for the retention period).

Would be good to
- keep old metric for transition period
- for new metrics add a simple prefix so that driver could easily ignore them when exporting to prometheus if it decides to
- add feature flag that allows reducing cardinality back (disable extended metrics completely), increase could be causing too big change in load towards metrics store, potentially, would be good to have a quick switch off
- maybe add feature flags per dimension to disable the breakdown: partition (very deep debugging thing and big multiplier at the same time, first candidate for being disabled), maybe others

Also maybe it is still to difficult to correctly discard metrics for failed/lost and retried data reads/writes, maybe there's some better way to get number of items read/write. Otoh metrics cleanup for completed jobs is a pretty clear improvement which would be nice to get done sooner than retry-aware counting (and so independently of it as it is proving to be not that trivial to get those numbers..)

Reply:
Following some discussion I'll need to read into what exactly happens with those metrics :eyes: plus when the driver side is ready would be easier to review them together :+1:

Reply:
* The new driver is backwards compatible, so we can bring this change in an out
* The driver will only send an extra label to prometheus: `completed`

The driver PR is ready: https://github.com/cognitedata/jetfire-backend/pull/2285